rfcs/2021-07-27-7117-vrl-string-interpolation.md

```diff
@@ -0,0 +1,197 @@
+```
+i'The message is {{ .message }} created at {{ .timestamp: %v %R }}'
+```
+
+and create an AST identical to the AST for the following expression:
+
+```
+s'The message is ' +
+as_string(.message) +
+s' created at ' +
+as_string(.timestamp, format: "%v %R")
+```
+
+`as_string` is a new function that will convert any type to a string and potentially
+apply format strings. Objects will be json encoded.
+
+## Rationale
+
+String interpolation or string formats are prevalent in modern programming
+languages. Users have an expectation that this feature will be available.
+
+String formatting is a common task within VRL. Currently the process involves
+string concatenation. This works, but the code required to do this does not
+create an immediately apparent representation of what the string may look like.
+
+There is little impact of not doing this beyond requiring users to use a less
+elegant form for string creation.
+
+## Prior Art
+
+- Template strings are used within certain fields within Vector.
+- Many programming languages offer string interpolation.
+  - [Python](https://datagy.io/python-f-strings/)
+  - [Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)
+  - [Ruby](http://ruby-for-beginners.rubymonstas.org/bonus/string_interpolation.html)
+
+## Drawbacks
+
+This does add additional code complexity and an maintenance burden to VRL.
+
+## Alternatives
+
+### Format Strings
+
+One alternative is to create a `sprintf` function. `sprintf` takes one parameter
+that represents the format string. Within the format string if a format tag is
+found, the function will take the next parameter passed and will format that
+parameter according to the tag and will embed the resulting text in that
+position.
+
+For example:
+
+```
+sprintf("The message is %s created at %t", .message, .timestamp)
+```
+
+will return
+
+```
+The message is the message created at Tue, 27 Jul 2021 10:10:01 +0000
+```
+
+The advantages of this method is that it does not require any changes to the
+VRL compiler, all changes are isolated to a single function. Also it provides
+a way to influence the formatting of the parameters.
+
+Downsides are that the format strings are a hidden DSL themselves and there is
+a cognitive overhead involved in maintaining the position of the format tags
+within the string and the parameters passed to the function.
+
+### Extended template strings
+
+The advantage of format strings over string interpolation is that format strings
+provide for parameters to indicate how numbers could be formatted.
+
+For example, `sprintf("%.2f", 3.14159)` will return `3.14`. With simple string
+interpolation `i'{{ 3.14159 }}'` there is no way to control the number of
+decimals output.
+
+We could extend the format strings to allow for any format parameters to be
+added after a `|` character in the template string. So
+`i'{{ 3.14159 | decimals: 2 }}'` would result in the number formatted to 2
+decimal places - `3.14`.
```

Comment:
This feels like it'd be a lot of new config/code to add, as each type would have it's own options here - right? Not that it's much different than having to add support for `%.2f`